Enh/Add hatch pattern support to Axes.grouped_bar

ilakkmanoharan opened the pull request:
Closes https://github.com/matplotlib/matplotlib/issues/30673

## PR summary
This PR enhances the provisional Axes.grouped_bar() API by introducing per-dataset hatch pattern support

Supersedes #30683 — rebased cleanly to resolve CI pr_clean conflict.

## Summary of Changes

Added a new hatch keyword argument to Axes.grouped_bar().
Updated the docstring to document hatch behavior and ValueError conditions.
Integrated hatch patterns into the plotting logic for both orientations:
Passed hatch=hatch_pattern to each call of .bar() and .barh().
Implemented validation and cycling behavior consistent with colors:
If hatch=None, no hatching is applied.
If a sequence is provided (e.g., ['//', 'xx', '..']), patterns are cycled across datasets.
Single string inputs (e.g., '//') or non-iterable values raise a ValueError.
Ensured that hatches are applied consistently across vertical and horizontal orientations.

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -3190,6 +3190,19 @@ def grouped_bar(self, heights, *, positions=None, group_spacing=1.5, bar_spacing
+        ------
+        ValueError
+            If ``hatch`` is a single string or a non-iterable value.
+
```
```suggestion
```

We don't list ValueErrors for invalid inputs under `Raises`. That would be an arbitrary and potentially long list with little benefit. It's by convention in Python that invalid inputs raise ValueErrors.

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -3318,6 +3331,36 @@ def grouped_bar(self, heights, *, positions=None, group_spacing=1.5, bar_spacing
+
+            # TODO: Discussion —
+            #   Should grouped_bar() apply a default hatch pattern (e.g., '//')
+            #   when none is provided ?
```
```suggestion
```
No. We don't have any hatches by default anywhere.

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -3318,6 +3331,36 @@ def grouped_bar(self, heights, *, positions=None, group_spacing=1.5, bar_spacing
+            #               f"Expected {num_datasets} hatches, got {len(hatch)}"
+            #           )
+            #   This would enforce a strict 1:1 correspondence between
+            #   datasets and provided hatches, preventing silent cycling.
```
```suggestion
```
The issue tracker is the right place for this, not the code.

ilakkmanoharan commented:
When colors=None:

Matplotlib automatically cycles through the property color cycle (axes.prop_cycle), giving each dataset a distinct color by default.

When colors is an empty list ([]):

Treated as None, so the default property color cycle is used.

So I wanted to mimic the same behavior for hatch too

story645 commented:
> This is the first time I’m raising the question of whether hatch should also adopt a similar default (i.e., cycle through a predefined set of hatch patterns) when None or an empty list is provided.

No, the preference is for no hatch.

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -3318,6 +3328,33 @@ def grouped_bar(self, heights, *, positions=None, group_spacing=1.5, bar_spacing
+
+            if not hatch_list:
+                # Empty sequence → treat as no hatch.
+                hatches = itertools.cycle([None])
```
Let's raise instead. Semantically, `hatch=[]` does not make sense. Users should use `None` for "no hatch".

ilakkmanoharan replied:
That makes sense semantically — None is the canonical way to disable hatching.
However, in practice, there are scenarios where hatch can be constructed programmatically (e.g., through loops, Giles, or data pipelines). In such cases, it may naturally or unintentionally evaluate to an empty list — for example, via something like config.get("hatch", []). Raising an error in these situations could make otherwise valid automation brittle.
Would it be acceptable to issue a warning and treat an empty list as None instead, to preserve robustness? This would align with Matplotlib’s general philosophy of being tolerant and user-friendly toward benign input while still informing users of the preferred approach.


```
if not hatch_list:
    _api.warn_external(
        "'hatch' was an empty sequence; treating as no hatch. "
        "Use None explicitly to suppress this warning."
    )
    hatches = itertools.cycle([None])
```

timhoffm replied:
@ilakkmanoharan 

> plotting utilities may use defensive defaults like config.get("hatch", []) to avoid KeyError.

Just use `config.get("hatch")` instead.

> Similarly, dynamically generating hatches from metadata — e.g. [meta.get("hatch") for meta in datasets] — can yield an empty list when all datasets omit that field.

No. this returns a list with exactliy `N = len(datasets)` entries. When all datasets omit that field, the list has N `None` entries.

So my argument persists: There's no natural case  for an empty list.

ilakkmanoharan replied:
Even if these use cases may be rare or not widely observed at the moment, such scenarios could easily emerge in evolving data-driven or cloud-integrated visualization systems in the immediate future — often unexpectedly.

ilakkmanoharan replied:
Thanks, Tim — I appreciate the clear decision and your time in walking through the reasoning. Understood on keeping the API strict for now and prioritizing consistency with stackplot. I won’t push the point further, and I’ll make the required changes to the code soon.

ilakkmanoharan commented:
We do have PR cleanliness issue due to removing doc/api/next_api_changes/behavior/30726-IM.rst --> Do you suggest I create a new PR or is there a way to remove this PR cleanliness issue from your side? @timhoffm @story645

ilakkmanoharan commented:
> Can you rebase? Which IDE are you using? I can also rebase for you and push to this PR.

Rebased and removed PR cleanliness error. What are the next steps?

timhoffm commented:
Yes, please.

ilakkmanoharan commented:
> @timhoffm
> 
> [#30729 (comment)](https://github.com/matplotlib/matplotlib/issues/30729#issue-3588816807) --- I am seeing your comments here just now. Do you suggest that I handle this change too, in this pull request? If so, I will take care of it tomorrow.

@timhoffm @story645 

https://github.com/matplotlib/matplotlib/issues/30729#issue-3588816807

I’ve been thinking and decided to keep the common_kwargs refactor as a separate PR. This helps avoid scope creep in the current PR, which is focused on hatch support, and keeps the review simpler by separating behavioral changes from structural cleanup. It also improves visibility for future maintenance, since the shared-kwargs pattern could be applied more broadly later. I would like to open a new PR for this. Please advise regarding this.

timhoffm commented:
> keeps the review simpler by separating behavioral changes from structural cleanup.

This is generally a good strategy. However, the cleanup is tiny. It’s a three-line change, two of which are anyway touched by this PR.

So there’s no real disadvantage in doing it right away. But I also won’t object if you do separate PRs.

ilakkmanoharan commented:
> > keeps the review simpler by separating behavioral changes from structural cleanup.
> 
> This is generally a good strategy. However, the cleanup is tiny. It’s a three-line change, two of which are anyway touched by this PR.
> 
> So there’s no real disadvantage in doing it right away. But I also won’t object if you do separate PRs.

Thank you @timhoffm. I appreciate it. I would like to do a separate PR. The issue #30729 is currently closed. Can I reference this issue in the PR?

ilakkmanoharan commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -3318,6 +3327,38 @@ def grouped_bar(self, heights, *, positions=None, group_spacing=1.5, bar_spacing
+                # Sequence of hatch patterns: cycle through them as needed.
+                # Example: hatch=['//', 'xx', '..'] → patterns repeat across datasets.
+                hatches = itertools.cycle(hatch_list)
+
```
When hatch is given as a single string (e.g. "//"), we raise a ValueError. This prevents Matplotlib from incorrectly iterating over individual characters ('/', '/') instead of treating the hatch as a single pattern.

story645 replied:
> Specifically:
determine the number of datasets (num_datasets = len(heights)),
normalize hatch to one pattern per dataset, and
apply that hatch consistently to all patches for that dataset.

Sounds good to me, I think would also then allow for easier expansion into nesting (hatch per patch per dataset) if we wanted to go that direction.
